File: day11/code.R
library(tidyverse)
lines <- readLines("day11/input.txt")

initial_m <- t(sapply(
    lines,
    function(line){
        as.numeric(strsplit(line, "")[[1]])
    }
))
rownames(initial_m) <- NULL

nr = nrow(initial_m)
nc = nrow(initial_m)

index2rc <- function(index, nr){
    if(index %% nr == 0){
        index_r = nr
        index_c = index %/% nr
    }else{
        index_r = index %% nr
        index_c = index %/% nr + 1
    }
    c(index_r, index_c)
}

# part 1
m = initial_m
total_flashes = 0
for (step in 1:100){
    m = m + 1
    
    indexes = which(m > 9)
    flashed_index = indexes
    
    while(length(indexes) > 0){
        rc_list = lapply(indexes, index2rc, nr=nr)
        neighbor_indexes = do.call(
            "bind_rows",
            lapply(
                1:length(rc_list),
                function(j){
                    rc_index=rc_list[[j]]
                    expand_grid(
                        index_r = c(rc_index[1]-1,rc_index[1],rc_index[1]+1),
                        index_c = c(rc_index[2]-1,rc_index[2],rc_index[2]+1)
                    ) %>% 
                        filter(index_r >= 1, index_r <= nr, index_c >= 1, index_c <= nc) %>% 
                        filter(!(index_r==rc_index[1] & index_c==rc_index[2]))
                }
            )
        )
        # single_index = (neighbor_indexes$index_c - 1) * nr + neighbor_indexes$index_r
        # m[single_index] = m[single_index] + 1
        
        for (i in 1:nrow(neighbor_indexes)){
            index_r = neighbor_indexes[i,"index_r"] %>% pull(index_r)
            index_c = neighbor_indexes[i,"index_c"] %>% pull(index_c)
            m[index_r, index_c] = m[index_r, index_c] + 1
        }
        
        indexes = which(m > 9)[!which(m > 9) %in% flashed_index]
        flashed_index = which(m > 9)
    }
    total_flashes = total_flashes + length(flashed_index)
    m[flashed_index] = 0
}

total_flashes # 1725

# part 2
m = initial_m
step=0
while(sum(m) > 0){
    step = step + 1
    m = m + 1
    
    indexes = which(m > 9)
    flashed_index = indexes
    
    while(length(indexes) > 0){
        rc_list = lapply(indexes, index2rc, nr=nr)
        
        neighbor_indexes = do.call(
            "bind_rows",
            lapply(
                1:length(rc_list),
                function(j){
                    rc_index=rc_list[[j]]
                    expand_grid(
                        index_r = c(rc_index[1]-1,rc_index[1],rc_index[1]+1),
                        index_c = c(rc_index[2]-1,rc_index[2],rc_index[2]+1)
                    ) %>% 
                        filter(index_r >= 1, index_r <= nr, index_c >= 1, index_c <= nc) %>% 
                        filter(!(index_r==rc_index[1] & index_c==rc_index[2]))
                }
            )
        )
        
        for (i in 1:nrow(neighbor_indexes)){
            index_r = neighbor_indexes[i,"index_r"] %>% pull(index_r)
            index_c = neighbor_indexes[i,"index_c"] %>% pull(index_c)
            m[index_r, index_c] = m[index_r, index_c] + 1
        }
        
        
        indexes = which(m > 9)[!which(m > 9) %in% flashed_index]
        flashed_index = which(m > 9)
    }
    m[flashed_index] = 0
}

step # 308

File: day11/code.py
#!/usr/bin/env python
# coding: utf-8

# In[122]:


import numpy as np


# In[123]:


with open("input.txt","r") as fh:
    lines = fh.readlines()
initial_m = [list(line.rstrip()) for line in lines]
initial_m = np.array(initial_m).astype(int)


# In[54]:


nr,nc = initial_m.shape


# In[115]:


m = initial_m
flashed_total = 0
for step in range(100):
    m = m + 1
    (index_r, index_c) = np.where(m > 9)
    flashed_indexes = [(index_r[x],index_c[x]) for x in np.arange(np.size(index_r))]
    flashed_indexes_total = flashed_indexes
    while (len(flashed_indexes) > 0):
        for i in np.arange(len(flashed_indexes)):
            neighbors_r = np.arange((flashed_indexes[i][0]-1),(flashed_indexes[i][0]+2))
            neighbors_c = np.arange((flashed_indexes[i][1]-1),(flashed_indexes[i][1]+2))
            neighbors_indexes_tmp = np.array([(r, c) for r in neighbors_r for c in neighbors_c if (r >=0 and r < nr and c >=0 and c < nc)])
            for j in np.arange(neighbors_indexes_tmp.shape[0]):
                m[neighbors_indexes_tmp[j,0],neighbors_indexes_tmp[j,1]] = m[neighbors_indexes_tmp[j,0],neighbors_indexes_tmp[j,1]] + 1
        (index_r, index_c) = np.where(m > 9)
        flashed_indexes_updated = [(index_r[t],index_c[t]) for t in np.arange(np.size(index_r))]        
        flashed_indexes = [index for index in flashed_indexes_updated if not(index in flashed_indexes_total)]
        flashed_indexes_total = flashed_indexes_updated
    m[np.where(m > 9)]= 0
    flashed_total += len(flashed_indexes_total)
flashed_total


# # part 2

# In[120]:


m = initial_m
step = 0
while(sum(sum(m)) > 0):
    step += 1
    m = m + 1
    (index_r, index_c) = np.where(m > 9)
    flashed_indexes = [(index_r[x],index_c[x]) for x in np.arange(np.size(index_r))]
    flashed_indexes_total = flashed_indexes
    while (len(flashed_indexes) > 0):
        for i in np.arange(len(flashed_indexes)):
            neighbors_r = np.arange((flashed_indexes[i][0]-1),(flashed_indexes[i][0]+2))
            neighbors_c = np.arange((flashed_indexes[i][1]-1),(flashed_indexes[i][1]+2))
            neighbors_indexes_tmp = [(r, c) for r in neighbors_r for c in neighbors_c if (r >=0 and r < nr and c >=0 and c < nc)]
            for j in np.arange(len(neighbors_indexes_tmp)):
                m[neighbors_indexes_tmp[j][0],neighbors_indexes_tmp[j][1]] = m[neighbors_indexes_tmp[j][0],neighbors_indexes_tmp[j][1]] + 1
        (index_r, index_c) = np.where(m > 9)
        flashed_indexes_updated = [(index_r[t],index_c[t]) for t in np.arange(np.size(index_r))]        
        flashed_indexes = [index for index in flashed_indexes_updated if not(index in flashed_indexes_total)]
        flashed_indexes_total = flashed_indexes_updated
    m[np.where(m > 9)]= 0

step

